core/src/main/java/org/elasticsearch/index/mapper/DocumentParser.java

```diff
@@ -442,31 +496,35 @@ private static ObjectMapper parseArray(ParseContext context, ObjectMapper parent
             } else if (dynamic == ObjectMapper.Dynamic.TRUE) {
                 Mapper.Builder builder = context.root().findTemplateBuilder(context, arrayFieldName, "object");
                 if (builder == null) {
-                    return parseNonDynamicArray(context, parentMapper, lastFieldName, arrayFieldName);
+                    // TODO: shouldn't this create a default object mapper builder?
```

Comment:
I don't think it should, we don't even know whether there is an object yet? I suspect the call to find a template builder above aims at giving templates the opportunity to return an `ArrayValueMapperParser` which would process the array all at once.